### ISSUE TITLE
Theme: Multithreading

### ISSUE THREAD

**Issue by SvyatoS1** [SvyatoS1/Course_preparation_C#16]

**Comment by SvyatoS1** [comment]
[04 - multithreading.pptx](https://github.com/SvyatoS1/Course_preparation_C/files/14271066/04.-.multithreading.pptx)

**Comment by PrymakBogdan** [comment]
В C++11 і новіших стандартах існує бібліотека `<thread>`, яка дозволяє просто створювати потоки. Ось найпростіший приклад створення потоку:

```cpp
#include <iostream>
#include <thread>

// Функція, яку буде виконувати потік
void myFunction() {
    std::cout << "Це виконується в потоці." << std::endl;
}

int main() {
    // Створення потоку та передача йому функції для виконання
    std::thread myThread(myFunction);

    // Очікування завершення виконання потоку
    myThread.join();

    // Головний потік продовжить виконання після завершення створеного потоку

    std::cout << "Це виконується в головному потоці." << std::endl;

    return 0;
}
```

У цьому прикладі функція `myFunction` викликається в окремому потоці, який був створений за допомогою `std::thread myThread(myFunction);`. Метод `join()` використовується для очікування завершення виконання потоку.

Зверніть увагу, що важливо викликати `join()` для кожного потоку перед завершенням програми, щоб гарантувати коректне завершення виконання потоків. У реальних програмах може бути бажано використовувати механізми керування потоками, такі як `std::async` або `std::future`, для кращого управління потоками та отримання результатів їх виконання.

**Comment by PrymakBogdan** [comment]
В C++ є кілька способів дочекатися на завершення виконання потоку. Один із засобів - використання методу `join()`, який чекає, доки потік завершить свою роботу. Однак, є й інші можливості, такі як `detach()`, `std::async`, і `std::future`. Розглянемо кілька з них:

1. **Використання `join()`:**
   ```cpp
   #include <iostream>
   #include <thread>

   void myFunction() {
       std::cout << "Це виконується в потоці." << std::endl;
   }

   int main() {
       std::thread myThread(myFunction);
       myThread.join(); // Очікування завершення виконання потоку
       std::cout << "Це виконується в головному потоці." << std::endl;

       return 0;
   }
   ```

2. **Використання `detach()`:**
   ```cpp
   #include <iostream>
   #include <thread>

   void myFunction() {
       std::cout << "Це виконується в потоці." << std::endl;
   }

   int main() {
       std::thread myThread(myFunction);
       myThread.detach(); // Відсоединення потоку (головний потік не буде чекати на його завершення)
       std::cout << "Це виконується в головному потоці." << std::endl;

       return 0;
   }
   ```

3. **Використання `std::async` та `std::future`:**
   ```cpp
   #include <iostream>
   #include <future>

   void myFunction() {
       std::cout << "Це виконується в потоці." << std::endl;
   }

   int main() {
       std::future<void> result = std::async(std::launch::async, myFunction);
       result.wait(); // Очікування завершення виконання потоку
       std::cout << "Це виконується в головному потоці." << std::endl;

       return 0;
   }
   ```

Вибір методу залежить від конкретних вимог вашого застосування. Обирайте `join()`, якщо потрібно дочекатися завершення виконання потоку перед продовженням роботи головного потоку. `detach()` може бути використаний, якщо ви бажаєте відсоединити потік, і головний потік не очікує на його завершення. І, нарешті, `std::async` і `std::future` можуть бути використані для отримання результату виконання або чекання завершення виконання потоку.

**Comment by PrymakBogdan** [comment]
Додавання таймауту до очікування завершення потоку можна реалізувати різними способами, залежно від конкретних вимог і можливостей. Наприклад, для методу `join()`, який блокує виконання головного потоку до завершення іншого, ви можете використовувати функцію `std::this_thread::sleep_for` для затримки виконання головного потоку протягом певного часу.

Ось приклад використання таймауту з `join()`:

```cpp
#include <iostream>
#include <thread>
#include <chrono>

void myFunction() {
    std::this_thread::sleep_for(std::chrono::seconds(3));
    std::cout << "Це виконується в потоці." << std::endl;
}

int main() {
    std::thread myThread(myFunction);

    // Очікування завершення виконання потоку протягом 2 секунд
    if (myThread.joinable()) {
        myThread.join();
    } else {
        std::cout << "Таймаут: Потік не був запущений або вже відсоединений." << std::endl;
    }

    std::cout << "Це виконується в головному потоці." << std::endl;

    return 0;
}
```

У цьому прикладі `std::this_thread::sleep_for(std::chrono::seconds(3));` служить як симуляція тривалості виконання функції у потоці.

Якщо вам потрібно здійснити очікування з таймаутом для `std::async` та `std::future`, ви можете використовувати `std::future_status` у поєднанні з `wait_for`:

```cpp
#include <iostream>
#include <future>
#include <chrono>

void myFunction() {
    std::this_thread::sleep_for(std::chrono::seconds(3));
    std::cout << "Це виконується в потоці." << std::endl;
}

int main() {
    std::future<void> result = std::async(std::launch::async, myFunction);

    // Очікування завершення виконання потоку протягом 2 секунд
    auto status = result.wait_for(std::chrono::seconds(2));

    if (status == std::future_status::ready) {
        std::cout << "Потік завершив виконання." << std::endl;
    } else if (status == std::future_status::timeout) {
        std::cout << "Таймаут: Потік не завершив виконання протягом вказаного часу." << std::endl;
    }

    std::cout << "Це виконується в головному потоці." << std::endl;

    return 0;
}
```

У цьому прикладі `result.wait_for(std::chrono::seconds(2))` чекає на завершення виконання потоку протягом 2 секунд.
